### PR TITLE
Implement Company::aiThink

### PR DESCRIPTION
**Pull request by ZehMatt**

### CODE REVIEW

**Review by AaronVanGeffen — COMMENTED** [review]
Apologies for the radio silence on testing this. I wonder what a good approach would be? There are calls to prng functions involved, and OpenLoco's prng implementation differs from vanilla. That rules out our usual batch of replay tests, unfortunately.

**Comment by ZehMatt** [comment]
> Apologies for the radio silence on testing this. I wonder what a good approach would be? There are calls to prng functions involved, and OpenLoco's prng implementation differs from vanilla. That rules out our usual batch of replay tests, unfortunately.

How/Why is the prng different? And if thats the case cant we make new ones with the new prng state.

**Comment by duncanspumpkin** [comment]
There are like 3 slight variations that cs used in locomotion. We only implemented one variation so if the code you implemented doesn't match that variation it will be different.

**Comment by ZehMatt** [comment]
> There are like 3 slight variations that cs used in locomotion. We only implemented one variation so if the code you implemented doesn't match that variation it will be different.

Would have been probably not a bad idea to have all 3 generator functions before turning them into one.

**Comment by duncanspumpkin** [comment]
This function is using the same gprng code as vanilla so it should match on a replay.

File: src/OpenLoco/Company.cpp [hunk 1]
```diff
@@ -23,9 +173,71 @@ namespace OpenLoco
...
+        const auto thinkFunc1 = _funcs_430786[var_4A4];
+        thinkFunc1(this);
+
+        if (empty())
+            return;
+
+        const auto thinkFunc2 = _funcs_43079E[var_4A6];
+        thinkFunc2(this);
+
+        if (headquarters_x == -1 && (challenge_flags & CompanyFlags::bankrupt) == 0 && (challenge_flags & 0x1) != 0)
```
Comment (duncanspumpkin) [inline src/OpenLoco/Company.cpp:191]: suggest inverting this and returning early.

Comment (duncanspumpkin) [inline src/OpenLoco/Company.cpp:191]: Add a defintion in the CompanyFlags table for (1<<0)

File: src/OpenLoco/Company.cpp [hunk 2]
```diff
@@ -23,9 +173,71 @@ namespace OpenLoco
...
+                pos = { town->x, town->y };
+            }
+
+            auto& prng = gPrng();
+            const auto randPick = prng.randNext();
+            const auto randPos = Map::Pos2{
+                static_cast<coord_t>(randPick & 0x3E0),
+                static_cast<coord_t>(Utility::ror<uint32_t>(randPick, 5) & 0x3E0)
+            };
+            const auto selectedPos = randPos - Map::Pos2{ 512, 512 } + pos;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Company.cpp:221]: I think making this `- Map::TilePos2{ 16, 16 }` would be more easy to read. Also Suggest moving the first sum into the randPos so it makes it obvious that the rand is +- 16 tiles.

File: src/OpenLoco/Company.cpp [hunk 3]
```diff
@@ -23,9 +173,71 @@ namespace OpenLoco
...
+            if (selectedPos.x <= 0x2FFF && selectedPos.y <= 0x2FFF)
```
Comment (duncanspumpkin) [inline src/OpenLoco/Company.cpp:222]: `Map::validCoords`

File: src/OpenLoco/Company.cpp [hunk 4]
```diff
@@ -23,9 +173,71 @@ namespace OpenLoco
...
+            {
+                auto tile = Map::TileManager::get(selectedPos);
+                auto* surface = tile.surface();
+
+                coord_t z = surface->baseZ() << 2;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Company.cpp:227]: * 4

File: src/OpenLoco/Company.cpp [hunk 5]
```diff
@@ -23,9 +173,71 @@ namespace OpenLoco
...
+                if (surface->slope() != 0)
+                    z += 16;
+
+                const auto rot = randPos.y & 3;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Company.cpp:231]: This wont work. randPos.y has been &0x3E0. You need to get the rand out of `randPick`.
rot = (Utility::ror<uint32_t>(randPick, 10)) & 0x3;

**Review by duncanspumpkin — CHANGES_REQUESTED** [review]
see comments

**Comment by duncanspumpkin** [comment]
No divergence after 1000 ticks

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**